### DISCUSSION TITLE
Error on “superset/utils/core.py”: No module named ‘cryptography.hazmat.backends.openssl.x509’

### DISCUSSION THREAD

**Discussion by actxact** [apache/superset#22613] [Q&A / Help] [answered]
Hi everyone.
I have a question on line 79 of the file “superset/utils/core.py”.
This one `from cryptography.hazmat.backends.openssl.x509 import _Certificate`

While executing `superset db upgrade` i got an error on this line - 
_ModuleNotFoundError: No module named ‘cryptography.hazmat.backends.openssl.x509’_

Because the latest version of “[Cryptography](https://github.com/pyca/cryptography/tree/main/src/cryptography/hazmat/backends/openssl)” does not contain such module.
This is a bug or am I doing something wrong?

**Comment by druce** [comment]
I've been struggling with the same issue. Dockerfile is below, I think I am doing exactly as described in the [installation instructions](https://superset.apache.org/docs/installation/installing-superset-from-scratch/)?
```
FROM ubuntu:20.04

# Install dependencies
RUN apt-get update && apt-get upgrade -y && \
  apt-get install -y build-essential libssl-dev libffi-dev python3-dev python3-pip libsasl2-dev libldap2-dev default-libmysqlclient-dev sudo

# Add user ubuntu with no password, add to sudo group
RUN adduser --disabled-password --gecos '' ubuntu && \
  adduser ubuntu sudo  && \
  echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers  && \
  echo 'ubuntu:ubuntu' | chpasswd && \ 
  echo 'root:password' | chpasswd

USER ubuntu
WORKDIR /home/ubuntu/
```
then in the running container, I do `sudo pip install --upgrade pip` , pip --version shows pip 22.3.1

then `sudo pip install apache-superset` completes with no error

then 
```
ubuntu@63b9e55fd20c:~$ superset db upgrade
Traceback (most recent call last):
  File "/usr/local/bin/superset", line 5, in <module>
    from superset.cli.main import superset
  File "/usr/local/lib/python3.8/dist-packages/superset/__init__.py", line 21, in <module>
    from superset.app import create_app
  File "/usr/local/lib/python3.8/dist-packages/superset/app.py", line 23, in <module>
    from superset.initialization import SupersetAppInitializer
  File "/usr/local/lib/python3.8/dist-packages/superset/initialization/__init__.py", line 33, in <module>
    from superset.extensions import (
  File "/usr/local/lib/python3.8/dist-packages/superset/extensions/__init__.py", line 32, in <module>
    from superset.utils.cache_manager import CacheManager
  File "/usr/local/lib/python3.8/dist-packages/superset/utils/cache_manager.py", line 24, in <module>
    from superset.utils.core import DatasourceType
  File "/usr/local/lib/python3.8/dist-packages/superset/utils/core.py", line 76, in <module>
    from cryptography.hazmat.backends.openssl.x509 import _Certificate
ModuleNotFoundError: No module named 'cryptography.hazmat.backends.openssl.x509'
```
any suggestion much appreciated!

**Reply by actxact** [reply]
i solved the problem
you need to roll back the version of the module Cryptography
`pip install --force-reinstall cryptography==38.0.4`

and another module requires version rollback - WTForms
`pip install --force-reinstall wtforms==2.3.3`

**Comment by sfirke** [comment, marked as answer]
See also discussion at https://github.com/apache/superset/issues/22571

**Comment by druce** [comment]
that worked for me, many thanks!

in order to export dashboards I also needed
pip install --force-reinstall flask==2.1

otherwise I got:
typeerror: send_file() got an unexpected keyword argument 'cache_timeout'
